src/CTA.WebForms2Blazor/Services/WorkspaceManagerService.cs

```diff
@@ -0,0 +1,153 @@
+            }
+
+            var targetProject = _workspace.CurrentSolution.Projects.Where(project => project.Name.Equals(projectName)).Single();
+            var newSolution = _workspace.CurrentSolution.WithProjectReferences(targetProject.Id, projectReferences);
+
+            if (!_workspace.TryApplyChanges(newSolution))
+            {
+                throw new InvalidOperationException(string.Format(WORKSPACE_UPDATE_FAILED_ERROR_TEMPLATE, "add project reference"));
+            }
+        }
+
+        public void AddMetadataReferences(string projectName, IEnumerable<MetadataReference> metadataReferences)
+        {
+            if (_workspace == null)
+            {
+                throw new InvalidOperationException(string.Format(WORKSPACE_MISSING_ERROR_TEMPLATE, "add metadata reference"));
+            }
+
+            var targetProject = _workspace.CurrentSolution.Projects.Where(project => project.Name.Equals(projectName)).Single();
+            var newSolution = _workspace.CurrentSolution.WithProjectMetadataReferences(targetProject.Id, metadataReferences);
+
+            if (!_workspace.TryApplyChanges(newSolution))
+            {
+                throw new InvalidOperationException(string.Format(WORKSPACE_UPDATE_FAILED_ERROR_TEMPLATE, "add metadata reference"));
+            }
+        }
+
+        public Document AddDocument(string projectName, string documentName, string documentText)
+        {
+            if (_workspace == null)
+            {
+                throw new InvalidOperationException(string.Format(WORKSPACE_MISSING_ERROR_TEMPLATE, "add document"));
+            }
+
+            var targetProject = _workspace.CurrentSolution.Projects.Where(project => project.Name.Equals(projectName)).Single();
+            Document document = _workspace.AddDocument(targetProject.Id, documentName, SourceText.From(documentText));
+            _numDocuments += 1;
+
+            return document;
+        }
+
+        public void NotifyNewExpectedProject()
+        {
+            _expectedProjects += 1;
+        }
+
+        public void NotifyNewExpectedDocument()
+        {
+            _expectedDocuments += 1;
+        }
+
+        public async Task WaitUntilAllProjectsInWorkspace(CancellationToken token)
+        {
+            while (_numProjects < _expectedProjects)
+            {
+                if (token.IsCancellationRequested)
+                {
+                    throw new OperationCanceledException(token);
+                }
+
+                await Task.Delay(25);
+            }
+        }
+
+        public async Task WaitUntilAllDocumentsInWorkspace(CancellationToken token)
+        {
+            while (_numDocuments < _expectedDocuments)
+            {
+                if (token.IsCancellationRequested)
+                {
+                    throw new OperationCanceledException(token);
+                }
+
+                await Task.Delay(25);
+            }
+        }
```

Comment:
Is there a reason for the explicit delays? Or is it just a placeholder so the compiler doesn't complain about the methods being `async` with no `await`?

Reply:
These two functions exist for use by FileInformation objects or ClassInformation objects so that they can pause their execution if a critical condition isn't met yet. For instance, we want to hold off on adding documents to the workspace and hold off on starting code migrations until the solution and projects have been added to both workspaces. Not sure if there's a better way to do it outside of Task.Delay in a loop that checks the condition though

Reply:
Got it, I understand now. I’m not sure of a better way at the moment but we can revisit later if we need to. Should be ok for now

Reply:
We could end up with infinite loop here if somehow condition is never met, can we have an escape condition if this runs for too long?